kubernetes-csi/external-snapshotter
# DeleteVolumeGroupSnapshot API - part 1

**What type of PR is this?**
/kind feature


**What this PR does / why we need it**:

This PR adds logic to delete volume group snapshots. 

Currently, Only the volume group snapshots and corresponding volume group snapshot contents are deleted. The logic to delete individual snapshots will be added in a separate PR.

**Testing**

1. Dynamically volume group snapshot:

- Create a volume group snapshot 
```
% kubectl create -f volumegroupsnapshot-example.yaml 
volumegroupsnapshot.groupsnapshot.storage.k8s.io/my-group-snapshot created
% kubectl get vgsc
NAME                                                    READYTOUSE   DELETIONPOLICY   DRIVER                VOLUMEGROUPSNAPSHOTCLASS   VOLUMEGROUPSNAPSHOTNAMESPACE   VOLUMEGROUPSNAPSHOT   AGE
groupsnapcontent-29f97dbc-d2a7-4204-b356-98a8c29870fb   true         Delete           hostpath.csi.k8s.io   vgs-class                  default                        my-group-snapshot     10s
% kubectl get vgs 
NAME                READYTOUSE   VOLUMEGROUPSNAPSHOTCLASS   VOLUMEGROUPSNAPSHOTCONTENT                              CREATIONTIME   AGE
my-group-snapshot   true         vgs-class                  groupsnapcontent-29f97dbc-d2a7-4204-b356-98a8c29870fb   11s            11s
% kubectl get vs  
NAME                                                                                           READYTOUSE   SOURCEPVC   SOURCESNAPSHOTCONTENT                                                                             RESTORESIZE   SNAPSHOTCLASS   SNAPSHOTCONTENT                                                                                   CREATIONTIME   AGE
snapshot-0a89a20f6f519bbb505bd29e6e8651e50319d3e527705239f9687d1fbce1a8a6-2023-10-13-8.50.54   true                     snapcontent-0a89a20f6f519bbb505bd29e6e8651e50319d3e527705239f9687d1fbce1a8a6-2023-10-13-8.50.54   1Gi                           snapcontent-0a89a20f6f519bbb505bd29e6e8651e50319d3e527705239f9687d1fbce1a8a6-2023-10-13-8.50.54   19s            19s
snapshot-4c6d03c90c122b50363288a808f0e784469b9730117fa028df007967a6b19359-2023-10-13-8.50.54   true                     snapcontent-4c6d03c90c122b50363288a808f0e784469b9730117fa028df007967a6b19359-2023-10-13-8.50.54   1Gi                           snapcontent-4c6d03c90c122b50363288a808f0e784469b9730117fa028df007967a6b19359-2023-10-13-8.50.54   19s            19s
% kubectl get vsc
NAME                                                                                              READYTOUSE   RESTORESIZE   DELETIONPOLICY   DRIVER                VOLUMESNAPSHOTCLASS   VOLUMESNAPSHOT                                                                                 VOLUMESNAPSHOTNAMESPACE   AGE
snapcontent-0a89a20f6f519bbb505bd29e6e8651e50319d3e527705239f9687d1fbce1a8a6-2023-10-13-8.50.54   true         1073741824    Delete           hostpath.csi.k8s.io                         snapshot-0a89a20f6f519bbb505bd29e6e8651e50319d3e527705239f9687d1fbce1a8a6-2023-10-13-8.50.54   default                   20s
snapcontent-4c6d03c90c122b50363288a808f0e784469b9730117fa028df007967a6b19359-2023-10-13-8.50.54   true         1073741824    Delete           hostpath.csi.k8s.io                         snapshot-4c6d03c90c122b50363288a808f0e784469b9730117fa028df007967a6b19359-2023-10-13-8.50.54   default                   20s
```

- Delete the volume group snapshot and verify that objects are deleted from API server:
```
% kubectl delete -f volumegroupsnapshot-example.yaml 
volumegroupsnapshot.groupsnapshot.storage.k8s.io "my-group-snapshot" deleted
% kubectl get vgs
No resources found in default namespace.
% kubectl get vgsc
No resources found
```

- Verify that VS and VSC still exist:
```
% kubectl get vs  
NAME                                                                                           READYTOUSE   SOURCEPVC   SOURCESNAPSHOTCONTENT                                                                             RESTORESIZE   SNAPSHOTCLASS   SNAPSHOTCONTENT                                                                                   CREATIONTIME   AGE
snapshot-0a89a20f6f519bbb505bd29e6e8651e50319d3e527705239f9687d1fbce1a8a6-2023-10-13-8.50.54   true                     snapcontent-0a89a20f6f519bbb505bd29e6e8651e50319d3e527705239f9687d1fbce1a8a6-2023-10-13-8.50.54   1Gi                           snapcontent-0a89a20f6f519bbb505bd29e6e8651e50319d3e527705239f9687d1fbce1a8a6-2023-10-13-8.50.54   19s            19s
snapshot-4c6d03c90c122b50363288a808f0e784469b9730117fa028df007967a6b19359-2023-10-13-8.50.54   true                     snapcontent-4c6d03c90c122b50363288a808f0e784469b9730117fa028df007967a6b19359-2023-10-13-8.50.54   1Gi                           snapcontent-4c6d03c90c122b50363288a808f0e784469b9730117fa028df007967a6b19359-2023-10-13-8.50.54   19s            19s
% kubectl get vsc
NAME                                                                                              READYTOUSE   RESTORESIZE   DELETIONPOLICY   DRIVER                VOLUMESNAPSHOTCLASS   VOLUMESNAPSHOT                                                                                 VOLUMESNAPSHOTNAMESPACE   AGE
snapcontent-0a89a20f6f519bbb505bd29e6e8651e50319d3e527705239f9687d1fbce1a8a6-2023-10-13-8.50.54   true         1073741824    Delete           hostpath.csi.k8s.io                         snapshot-0a89a20f6f519bbb505bd29e6e8651e50319d3e527705239f9687d1fbce1a8a6-2023-10-13-8.50.54   default                   20s
snapcontent-4c6d03c90c122b50363288a808f0e784469b9730117fa028df007967a6b19359-2023-10-13-8.50.54   true         1073741824    Delete           hostpath.csi.k8s.io                         snapshot-4c6d03c90c122b50363288a808f0e784469b9730117fa028df007967a6b19359-2023-10-13-8.50.54   default                   20s
```


**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
Introduce logic to delete volume group snapshots
```
